File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/library/artists/items/ArtistItemsPresenter.java
package com.github.anrimian.musicplayer.ui.library.artists.items;

import static com.github.anrimian.musicplayer.data.utils.rx.RxUtils.dispose;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.anrimian.musicplayer.data.utils.rx.RxUtils;
import com.github.anrimian.musicplayer.domain.interactors.library.LibraryArtistsInteractor;
import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor;
import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor;
import com.github.anrimian.musicplayer.domain.interactors.settings.DisplaySettingsInteractor;
import com.github.anrimian.musicplayer.domain.models.albums.Album;
import com.github.anrimian.musicplayer.domain.models.artist.Artist;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;
import com.github.anrimian.musicplayer.ui.library.common.compositions.BaseLibraryCompositionsPresenter;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.disposables.Disposable;


public class ArtistItemsPresenter extends BaseLibraryCompositionsPresenter<ArtistItemsView> {

    private final long artistId;
    private final LibraryArtistsInteractor interactor;

    private Disposable changeDisposable;

    private Artist artist;

    @Nullable
    private Completable lastEditAction;

    public ArtistItemsPresenter(long artistId,
                                LibraryArtistsInteractor interactor,
                                PlayListsInteractor playListsInteractor,
                                LibraryPlayerInteractor playerInteractor,
                                DisplaySettingsInteractor displaySettingsInteractor,
                                ErrorParser errorParser,
                                Scheduler uiScheduler) {
        super(playerInteractor,
                playListsInteractor,
                displaySettingsInteractor,
                errorParser,
                uiScheduler);
        this.artistId = artistId;
        this.interactor = interactor;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        subscribeOnArtistInfo();
        subscribeOnArtistAlbums();
    }

    @NonNull
    @Override
    protected Observable<List<Composition>> getCompositionsObservable(String searchText) {
        return interactor.getCompositionsByArtist(artistId);
    }

    @Override
    protected ListPosition getSavedListPosition() {
        return interactor.getSavedItemsListPosition(artistId);
    }

    @Override
    protected void saveListPosition(@NonNull ListPosition listPosition) {
        interactor.saveItemsListPosition(artistId, listPosition);
    }

    void onFragmentMovedToTop() {
        //save selected screen. Wait a little for all screens
        if (artist != null) {
            getViewState().showArtistInfo(artist);
        }
    }

    void onRenameArtistClicked() {
        if (artist != null) {
            getViewState().showRenameArtistDialog(artist);
        }
    }

    void onNewArtistNameEntered(String name, long artistId) {
        dispose(changeDisposable);
        lastEditAction = interactor.updateArtistName(name, artistId)
                .observeOn(getUiScheduler())
                .doOnSubscribe(d -> getViewState().showRenameProgress())
                .doFinally(() -> getViewState().hideRenameProgress());
        changeDisposable = lastEditAction.subscribe(() -> {}, this::onDefaultError);
    }

    void onRetryFailedEditActionClicked() {
        if (lastEditAction != null) {
            RxUtils.dispose(changeDisposable, getPresenterDisposable());
            changeDisposable = lastEditAction
                    .doFinally(() -> lastEditAction = null)
                    .subscribe(() -> {}, this::onDefaultError);
            getPresenterDisposable().add(changeDisposable);
        }
    }

    private void subscribeOnArtistInfo() {
        getPresenterDisposable().add(interactor.getArtistObservable(artistId)
                .observeOn(getUiScheduler())
                .subscribe(this::onArtistInfoReceived,
                        t -> getViewState().closeScreen(),
                        getViewState()::closeScreen));
    }

    private void onArtistInfoReceived(Artist artist) {
        this.artist = artist;
        getViewState().showArtistInfo(artist);
    }

    private void subscribeOnArtistAlbums() {
        getPresenterDisposable().add(interactor.getAllAlbumsForArtist(artistId)
                .observeOn(getUiScheduler())
                .subscribe(this::onArtistInfoReceived,
                        t -> getViewState().closeScreen(),
                        getViewState()::closeScreen));
    }

    private void onArtistInfoReceived(List<Album> albums) {
        getViewState().showArtistAlbums(albums);
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/library/artists/items/ArtistItemsPresenter.kt
package com.github.anrimian.musicplayer.ui.library.artists.items

import com.github.anrimian.filesync.SyncInteractor
import com.github.anrimian.musicplayer.domain.interactors.library.LibraryArtistsInteractor
import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor
import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.domain.interactors.settings.DisplaySettingsInteractor
import com.github.anrimian.musicplayer.domain.models.artist.Artist
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.sync.FileKey
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.library.common.compositions.BaseLibraryCompositionsPresenter
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Scheduler

class ArtistItemsPresenter(
    private val artistId: Long,
    private val interactor: LibraryArtistsInteractor,
    playerInteractor: LibraryPlayerInteractor,
    displaySettingsInteractor: DisplaySettingsInteractor,
    syncInteractor: SyncInteractor<FileKey, *, Long>,
    playListsInteractor: PlayListsInteractor,
    errorParser: ErrorParser,
    uiScheduler: Scheduler
) : BaseLibraryCompositionsPresenter<Composition, ArtistItemsView>(
    displaySettingsInteractor,
    syncInteractor,
    playerInteractor,
    playListsInteractor,
    errorParser,
    uiScheduler
) {

    private var artist: Artist? = null

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        subscribeOnArtistInfo()
        subscribeOnArtistAlbums()
    }

    override fun getCompositionsObservable(searchText: String?): Observable<List<Composition>> {
        return interactor.getCompositionsByArtist(artistId)
    }

    override fun getSavedListPosition(): ListPosition? {
        return interactor.getSavedItemsListPosition(artistId)
    }

    override fun saveListPosition(listPosition: ListPosition) {
        interactor.saveItemsListPosition(artistId, listPosition)
    }

    fun onFragmentResumed() {
        interactor.setSelectedArtistScreen(artistId)
        if (artist != null) {
            viewState.showArtistInfo(artist!!)
        }
    }

    fun onRenameArtistClicked() {
        if (artist != null) {
            viewState.showRenameArtistDialog(artist!!)
        }
    }

    private fun subscribeOnArtistInfo() {
        interactor.getArtistObservable(artistId).subscribeOnUi(
            this::onArtistInfoReceived,
            { viewState.closeScreen() },
            viewState::closeScreen
        )
    }

    private fun onArtistInfoReceived(artist: Artist) {
        this.artist = artist
        viewState.showArtistInfo(artist)
    }

    private fun subscribeOnArtistAlbums() {
        interactor.getAllAlbumsForArtist(artistId).subscribeOnUi(
            viewState::showArtistAlbums,
            { viewState.closeScreen() },
            viewState::closeScreen
        )
    }
}